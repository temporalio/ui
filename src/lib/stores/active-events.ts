import { get, writable } from 'svelte/store';

import type { EventGroup } from '$lib/models/event-groups/event-groups';
import type { WorkflowEvent } from '$lib/types/events';

export const indexPageSize = 200;
export const startIndex = writable(0);
export const endIndex = writable(indexPageSize);

export const activeEvents = writable<string[]>([]);
export const activeGroups = writable<string[]>([]);

export const clearActives = () => {
  activeGroups.set([]);
  activeEvents.set([]);
  startIndex.set(0);
  endIndex.set(indexPageSize);
};

export const clearActiveEvents = () => {
  activeGroups.set([]);
  activeEvents.set([]);
};

export const setActiveGroup = (group: EventGroup) => {
  activeEvents.set([]);
  if (!get(activeGroups).includes(group.id)) {
    activeGroups.set([...get(activeGroups), group.id]);
  } else {
    activeGroups.set(get(activeGroups).filter((id) => id !== group.id));
  }
};

export const setSingleActiveGroup = (group: EventGroup) => {
  activeEvents.set([]);
  if (!get(activeGroups).includes(group.id)) {
    activeGroups.set([group.id]);
  } else {
    activeGroups.set([]);
  }
};

export const setSingleActiveEvent = (event: WorkflowEvent) => {
  activeGroups.set([]);
  if (!get(activeEvents).includes(event.id)) {
    activeEvents.set([event.id]);
  } else {
    activeEvents.set([]);
  }
};

export const setActiveEvent = (event: WorkflowEvent, group: EventGroup) => {
  activeGroups.set([]);

  if (!get(activeEvents).includes(event.id)) {
    const activeEventInGroup = get(activeEvents).find(
      (id) => group?.eventIds.has(id),
    );
    if (activeEventInGroup) {
      activeEvents.set([
        ...get(activeEvents).filter((id) => id !== activeEventInGroup),
        event.id,
      ]);
    } else {
      activeEvents.set([...get(activeEvents), event.id]);
    }
  } else {
    activeEvents.set(get(activeEvents).filter((id) => id !== event.id));
  }
};

export const setActiveGroupAndEvent = (
  group: EventGroup,
  event: WorkflowEvent,
) => {
  if (group) {
    if (!get(activeGroups).includes(group.id)) {
      activeGroups.set([...get(activeGroups), group.id]);
    } else {
      activeGroups.set(get(activeGroups).filter((id) => id !== group.id));
    }
  }

  if (event) {
    if (!get(activeEvents).includes(event.id)) {
      activeEvents.set([...get(activeEvents), event.id]);
    } else {
      activeEvents.set(get(activeGroups).filter((id) => id !== event.id));
    }
  }
};
