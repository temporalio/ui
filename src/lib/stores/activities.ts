import { derived, get, Readable } from 'svelte/store';
import {
  isActivityTaskScheduledEvent,
  isActivityTaskStartedEvent,
  isActivityTaskCompletedEvent,
  isActivityTaskFailedEvent,
  isActivityTaskTimedOutEvent,
  isActivityTaskCancelRequestedEvent,
  isActivityTaskCanceledEvent,
} from '$lib/utilities/is-event-type';

type ActivitiesMap = { [key: string]: Activity };
type ActivityStore = Readable<Activity[]> & {
  getActivity: (id: string | number) => Readable<Activity>;
};

const setActivityId = (activities: ActivitiesMap) => (id: string) => {
  if (!activities[id]) activities[id] = {};
  activities[id].id = id;
};

const setActivityStatus =
  (activities: ActivitiesMap) => (id: string, status: ActivityStatus) => {
    if (!activities[id]) activities[id] = {};
    activities[id].status = status;
  };

const addToActivities =
  (activities: ActivitiesMap) =>
  (
    id: string,
    property: keyof Omit<Activity, 'id' | 'status'>,
    event: HistoryEvent,
  ) => {
    if (!activities[id]) activities[id] = {};
    activities[id][property] = event;
  };

export const createActivityStore = <
  E extends HistoryEvent[],
  S extends Readable<E>,
>(
  eventStore: S,
): ActivityStore => {
  const store = derived(
    eventStore,
    (events) => {
      const activities: ActivitiesMap = {};
      const add = addToActivities(activities);
      const setId = setActivityId(activities);
      const setStatus = setActivityStatus(activities);

      for (const event of events) {
        if (isActivityTaskScheduledEvent(event)) {
          const id = event.activityTaskScheduledEventAttributes.activityId;
          add(id, 'activityTaskScheduledEvent', event);
          setId(id);
          setStatus(id, 'Started');
        }

        if (isActivityTaskStartedEvent(event)) {
          const id = String(
            event.activityTaskStartedEventAttributes.scheduledEventId,
          );
          add(id, 'activityTaskStartedEvent', event);
          setStatus(id, 'Started');
        }

        if (isActivityTaskCompletedEvent(event)) {
          const id = String(
            event.activityTaskCompletedEventAttributes.scheduledEventId,
          );
          add(id, 'activityTaskCompletedEvent', event);
          setStatus(id, 'Completed');
        }

        if (isActivityTaskFailedEvent(event)) {
          const id = String(
            event.activityTaskFailedEventAttributes.scheduledEventId,
          );
          add(id, 'activityTaskFailedEvent', event);
          setStatus(id, 'Failed');
        }

        if (isActivityTaskTimedOutEvent(event)) {
          const id = String(
            event.activityTaskTimedOutEventAttributes.scheduledEventId,
          );
          add(id, 'activityTaskTimedOutEvent', event);
          setStatus(id, 'TimedOut');
        }

        if (isActivityTaskCancelRequestedEvent(event)) {
          const id = String(
            event.activityTaskCancelRequestedEventAttributes.scheduledEventId,
          );
          add(id, 'activityTaskCancelRequestedEvent', event);
          setStatus(id, 'CancelRequested');
        }

        if (isActivityTaskCanceledEvent(event)) {
          const id = String(
            event.activityTaskCanceledEventAttributes.scheduledEventId,
          );
          add(id, 'activityTaskCanceledEvent', event);
          setStatus(id, 'Canceled');
        }
      }
      return activities;
    },
    {},
  );

  const activities = derived(store, ($store) => Object.values($store));

  return {
    ...activities,
    getActivity: (id) => derived(store, ($store) => get(store)[id] || {}),
  };
};
