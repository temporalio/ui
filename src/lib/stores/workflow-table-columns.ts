import { derived, type Readable } from 'svelte/store';
import { persistStore } from './persist-store';
import { customSearchAttributes } from './search-attributes';

export const MAX_PINNED_COLUMNS = 2;

export const WorkflowHeaderLabels = [
  'Status',
  'Workflow ID',
  'Run ID',
  'Type',
  'Start',
  'End',
  'History Size',
  'History Length',
  'Execution Time',
  'Execution Duration',
  'State Transitions',
  'Parent Namespace',
  'Parent Workflow ID',
  'Task Queue',
] as const;

export type WorkflowHeaderLabel = (typeof WorkflowHeaderLabels)[number];

// https://github.com/microsoft/TypeScript/issues/29729
// https://stackoverflow.com/a/61048124
// eslint-disable-next-line @typescript-eslint/ban-types
type AnyWorkflowHeaderLabel = WorkflowHeaderLabel | (string & {});

export type WorkflowHeader = {
  label: AnyWorkflowHeaderLabel;
  pinned: boolean;
};

type State = WorkflowHeader[];

type Action =
  | {
      type: 'WORKFLOW_COLUMN.ADD';
      payload: { label: AnyWorkflowHeaderLabel };
    }
  | {
      type: 'WORKFLOW_COLUMN.REMOVE';
      payload: { label: AnyWorkflowHeaderLabel };
    }
  | {
      type: 'WORKFLOW_COLUMN.PIN';
      payload: { label: AnyWorkflowHeaderLabel };
    }
  | { type: 'WORKFLOW_COLUMN.MOVE'; payload: { from: number; to: number } };

const DEFAULT_COLUMNS: WorkflowHeader[] = [
  { label: 'Status', pinned: true },
  { label: 'Workflow ID', pinned: true },
  { label: 'Run ID', pinned: false },
  { label: 'Type', pinned: false },
  { label: 'Start', pinned: false },
  { label: 'End', pinned: false },
];

const DEFAULT_AVAILABLE_COLUMNS: WorkflowHeader[] = [
  { label: 'History Size', pinned: false },
  { label: 'History Length', pinned: false },
  { label: 'Execution Time', pinned: false },
  { label: 'Execution Duration', pinned: false },
  { label: 'State Transitions', pinned: false },
  { label: 'Parent Namespace', pinned: false },
  { label: 'Parent Workflow ID', pinned: false },
  { label: 'Task Queue', pinned: false },
];

const workflowTableColumns = persistStore<WorkflowHeader[]>(
  'workflow-table-columns',
  DEFAULT_COLUMNS,
);

const pinnedColumnsWidth = persistStore<number>(
  'workflow-table-pinned-columns-width',
);

const availableWorkflowColumns: Readable<WorkflowHeader[]> = derived(
  [workflowTableColumns],
  ([$workflowTableColumns]) =>
    [...DEFAULT_COLUMNS, ...DEFAULT_AVAILABLE_COLUMNS].filter(
      (header) =>
        !$workflowTableColumns.some((column) => column.label === header.label),
    ),
);

const availableSearchAttributeColumns: Readable<WorkflowHeader[]> = derived(
  [customSearchAttributes, workflowTableColumns],
  ([$customSearchAttributes, $workflowTableColumns]) =>
    Object.keys($customSearchAttributes)
      .filter(
        (searchAttribute) =>
          !$workflowTableColumns.some(
            (column) => column.label === searchAttribute,
          ),
      )
      .map((key) => ({
        label: key,
        pinned: false,
      })),
);

const reducer = (action: Action, state: State): State => {
  switch (action.type) {
    case 'WORKFLOW_COLUMN.ADD': {
      const { label } = action.payload;

      return [...state, { label, pinned: false }];
    }
    case 'WORKFLOW_COLUMN.REMOVE': {
      const { label: labelToRemove } = action.payload;

      return state.filter(({ label }) => label !== labelToRemove);
    }
    case 'WORKFLOW_COLUMN.PIN': {
      const { label: labelToPin } = action.payload;
      const index = state.findIndex(({ label }) => label === labelToPin);

      const isPinned = state[index].pinned;

      let lastPinned = -1;
      for (let i = state.length - 1; i >= 0; i--) {
        if (state[i].pinned) {
          lastPinned = i;
          break;
        }
      }

      const newColumns = [...state];
      newColumns[index].pinned = !isPinned;

      if (index > lastPinned && !isPinned) {
        newColumns.splice(lastPinned + 1, 0, newColumns.splice(index, 1)[0]);
      } else if (index < lastPinned && isPinned) {
        newColumns.splice(lastPinned, 0, newColumns.splice(index, 1)[0]);
      }

      return newColumns;
    }
    case 'WORKFLOW_COLUMN.MOVE': {
      const { from, to } = action.payload;
      const isPinned = state[from].pinned;

      let lastPinned = 0;
      for (let i = state.length - 1; i >= 0; i--) {
        if (state[i].pinned) {
          lastPinned = i;
          break;
        }
      }

      const tempColumns = [...state];
      if (to <= lastPinned && !isPinned) {
        tempColumns[from].pinned = true;
      } else if (to > lastPinned && isPinned) {
        tempColumns[from].pinned = false;
      }

      tempColumns.splice(to, 0, tempColumns.splice(from, 1)[0]);

      return tempColumns.map((c, idx) =>
        idx > MAX_PINNED_COLUMNS - 1 ? { ...c, pinned: false } : c,
      );
    }
    default:
      return state;
  }
};

const dispatch = (action: Action) => {
  workflowTableColumns.update((previousState) =>
    reducer(action, previousState),
  );
};

const addColumn = (label: AnyWorkflowHeaderLabel) => {
  dispatch({ type: 'WORKFLOW_COLUMN.ADD', payload: { label } });
};

const removeColumn = (label: AnyWorkflowHeaderLabel) => {
  dispatch({ type: 'WORKFLOW_COLUMN.REMOVE', payload: { label } });
};

const moveColumn = (from: number, to: number) => {
  dispatch({ type: 'WORKFLOW_COLUMN.MOVE', payload: { from, to } });
};

const pinColumn = (label: AnyWorkflowHeaderLabel) => {
  dispatch({ type: 'WORKFLOW_COLUMN.PIN', payload: { label } });
};

export {
  workflowTableColumns,
  availableSearchAttributeColumns,
  availableWorkflowColumns,
  pinnedColumnsWidth,
  addColumn,
  removeColumn,
  moveColumn,
  pinColumn,
};
