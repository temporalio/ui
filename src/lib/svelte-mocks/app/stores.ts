import { readable } from 'svelte/store';

import type { Settings } from '$lib/types/global';

interface Page<Params extends Record<string, string> = Record<string, string>> {
  url: URL;
  params: Params;
  routeId: string | null;
  data: App.PageData;
  status: number;
  error: Error | null;
}

const settings: Settings = {
  auth: {
    enabled: false,
    options: null,
  },
  baseUrl: 'http://localhost:3000',
  codec: {
    endpoint: '',
    passAccessToken: false,
    includeCredentials: false,
  },
  defaultNamespace: 'default',
  disableWriteActions: false,
  showTemporalSystemNamespace: false,
  batchActionsDisabled: false,
  workflowResetDisabled: false,
  workflowCancelDisabled: false,
  workflowSignalDisabled: false,
  workflowTerminateDisabled: false,
  hideWorkflowQueryErrors: false,
  notifyOnNewVersion: true,
  feedbackURL: '',
  runtimeEnvironment: {
    isCloud: false,
    isLocal: true,
    envOverride: true,
  },
  version: '2.0.0',
};

const data: App.PageData = {
  settings,
};

export const page = readable<Page>({
  error: null,
  params: {
    namespace: 'default',
  },
  routeId: 'namespaces/[namespace]/workflows@root',
  status: 200,
  data,
  url: new URL(
    'http://localhost:3000/namespaces/default/workflows?search=basic&query=WorkflowType%3D%22testing%22',
  ),
});
