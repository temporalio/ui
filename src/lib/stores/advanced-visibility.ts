import { derived } from 'svelte/store';

import { page } from '$app/stores';

import {
  advancedVisibilityEnabled,
  advancedVisibilityEnabledWithOrderBy,
} from '$lib/utilities/advanced-visibility-enabled';

import { cluster } from './cluster';
import { temporalVersion } from './versions';

export const isCloud = derived(
  [page],
  ([$page]) => $page.data?.settings?.runtimeEnvironment?.isCloud,
);

export const supportsAdvancedVisibility = derived(
  [cluster, temporalVersion, isCloud],
  ([$cluster, $temporalVersion, $isCloud]) =>
    advancedVisibilityEnabled($cluster, $temporalVersion) || $isCloud,
);

export const supportsAdvancedVisibilityWithOrderBy = derived(
  [cluster, isCloud],
  ([$cluster, $isCloud]) =>
    advancedVisibilityEnabledWithOrderBy($cluster) || $isCloud,
);
