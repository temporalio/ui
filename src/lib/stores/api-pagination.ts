import type { Readable } from 'svelte/store';
import { derived, writable } from 'svelte/store';

import { page } from '$app/stores';

import {
  defaultItemsPerPage,
  options,
  perPageFromSearchParameter,
  perPageKey,
} from './pagination';

type PaginationMethods<T> = {
  nextPageWithItems: (t: string, items: T[]) => void;
  nextPage: () => void;
  previousPage: () => void;
  setUpdating: () => void;
  reset: () => void;
  resetPageSize: (pageSize: number) => void;
  nextRow: () => void;
  previousRow: () => void;
  setActiveIndex: (activeIndex: number) => void;
};

type PaginationItems<T> = {
  key: string;
  loading: boolean;
  updating: boolean;
  hasNext: boolean;
  hasPrevious: boolean;
  index: number;
  previousPageSize: number;
  pageSize: number;
  indexData: Record<
    number,
    { nextToken: string; start: number; end: number; items: T[] }
  >;
  visibleItems: T[];
  hasNextIndexData: boolean;
  indexStart: number;
  indexEnd: number;
  activeIndex: number;
};

const defaultStore = {
  key: perPageKey,
  loading: true,
  updating: false,
  hasNext: false,
  hasPrevious: false,
  index: 0,
  previousPageSize: defaultItemsPerPage,
  pageSize: defaultItemsPerPage,
  indexData: {},
  visibleItems: [],
  hasNextIndexData: false,
  indexStart: 0,
  indexEnd: 0,
  activeIndex: 0,
};

export type PaginationStore<T> = PaginationMethods<T> &
  Readable<PaginationItems<T>>;

const setFirstOption = (options: string[] | number[]) => {
  const defaultOption = options[0];
  if (!defaultOption) return defaultItemsPerPage;
  const optionAsInt = parseInt(defaultOption.toString());
  if (isNaN(optionAsInt)) return defaultItemsPerPage;
  return optionAsInt;
};

export const getInitialPageSize = (
  options: string[] | number[],
  defaultPageSize: string | number | undefined = undefined,
) => {
  if (defaultPageSize) {
    const optionAsInt = parseInt(defaultPageSize.toString());
    if (isNaN(optionAsInt)) return setFirstOption(options);
    return optionAsInt;
  } else {
    return setFirstOption(options);
  }
};

export function createPaginationStore<T>(
  pageSizeOptions: string[] | number[] = options,
  defaultPageSize: string | number | undefined = undefined,
): PaginationStore<T> {
  const initialPageSize = getInitialPageSize(pageSizeOptions, defaultPageSize);
  const paginationStore = writable({
    ...defaultStore,
    previousPageSize: initialPageSize,
    pageSize: initialPageSize,
  });
  const { set, update } = paginationStore;

  const pageSize = derived([page], ([$page]) => {
    const perPage = $page.url.searchParams.get(perPageKey);
    return perPage ? perPageFromSearchParameter(perPage) : undefined;
  });

  const { subscribe } = derived(
    [paginationStore, pageSize],
    ([$pagination, $pageSize]) => {
      const visibleItems =
        $pagination.indexData[$pagination.index]?.items ?? [];
      const indexStart = $pagination.indexData[$pagination.index]?.start ?? 0;
      const indexEnd = $pagination.indexData[$pagination.index]?.end ?? 0;
      const hasNextIndexData = Boolean(
        $pagination.indexData[$pagination.index + 1],
      );

      return {
        ...$pagination,
        visibleItems,
        indexStart,
        indexEnd,
        hasNextIndexData,
        pageSize: $pageSize ?? $pagination.pageSize,
      };
    },
  );

  const setNextPageWithItems = <T>(
    nextToken: string,
    items: T[],
    _store: PaginationItems<T>,
  ) => {
    const currentIndex = _store.index;
    const store = {
      ..._store,
      indexData: { ..._store.indexData },
      hasNext: Boolean(nextToken),
      updating: false,
      loading: false,
    };

    if (!items.length) return { ...store, hasNext: false };

    if (!store.indexData[currentIndex]) {
      store.indexData[currentIndex] = {
        nextToken,
        start: 1,
        end: items.length,
        items,
      };
      store.hasPrevious = false;
    } else {
      (store.index = currentIndex + 1),
        (store.indexData[store.index] = {
          nextToken,
          start: store.indexData[currentIndex].end + 1,
          end: store.indexData[currentIndex].end + items.length,
          items,
        });
      store.hasPrevious = true;
    }

    return store;
  };

  const setNextPage = <T>(_store: PaginationItems<T>) => {
    const store = {
      ..._store,
      index: _store.index + 1,
      hasPrevious: true,
      loading: false,
      updating: false,
    };

    if (!store.indexData[store.index]?.nextToken) {
      store.hasNext = false;
    }

    return store;
  };

  const setPreviousPage = <T>(_store: PaginationItems<T>) => {
    const store = {
      ..._store,
      hasNext: true,
      updating: false,
      loading: false,
      index: _store.index > 0 ? _store.index - 1 : 0,
    };

    if (store.index === 0) {
      store.hasPrevious = false;
    }

    return store;
  };

  const setNextRow = <T>(_store: PaginationItems<T>) => {
    const store = { ..._store };
    const indexLength = store.indexData[store.index]?.items?.length ?? 0;

    if (store.activeIndex < indexLength - 1) {
      store.activeIndex = store.activeIndex + 1;
    }

    return store;
  };

  const setPreviousRow = <T>(_store: PaginationItems<T>) => {
    const store = { ..._store };
    const activeIndex = store.activeIndex >= 1 ? store.activeIndex - 1 : 0;

    store.activeIndex = activeIndex;

    return store;
  };

  const setActiveIndex = <T>(
    _store: PaginationItems<T>,
    activeIndex: number,
  ) => {
    return { ..._store, activeIndex };
  };

  const resetPageSize = <T>(_store: PaginationItems<T>, pageSize: number) => {
    return {
      ..._store,
      pageSize,
      previousPageSize: pageSize,
      index: 0,
      indexData: {},
      loading: true,
      updating: false,
    };
  };

  return {
    subscribe,
    nextPageWithItems: (token: string, items: T[]) =>
      update((store) => setNextPageWithItems(token, items, store)),
    nextPage: () => update((store) => setNextPage(store)),
    previousPage: () => update((store) => setPreviousPage(store)),
    setUpdating: () => update((store) => ({ ...store, updating: true })),
    reset: () => set(defaultStore),
    resetPageSize: (pageSize) =>
      update((store) => resetPageSize(store, pageSize)),
    nextRow: () => update((store) => setNextRow(store)),
    previousRow: () => update((store) => setPreviousRow(store)),
    setActiveIndex: (activeIndex: number) =>
      update((store) => setActiveIndex(store, activeIndex)),
  };
}
