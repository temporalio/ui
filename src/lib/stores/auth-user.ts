import { get } from 'svelte/store';

import { persistStore } from '$lib/stores/persist-store';
import type { User } from '$lib/types/global';

export const authUser = persistStore<User>('AuthUser', {});

export const getAuthUser = (): User => get(authUser);

export const setAuthUser = (user: User) => {
  const { accessToken, idToken, name, email, picture } = user;

  if (!accessToken) {
    throw new Error('No access token');
  }

  authUser.set({
    accessToken,
    idToken,
    name,
    email,
    picture,
  });
};

export const clearAuthUser = () => {
  authUser.set({});
};
