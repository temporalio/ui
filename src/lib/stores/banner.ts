import { get } from 'svelte/store';

import { persistStore } from '$lib/stores/persist-store';

export const closedBanners = persistStore<string[]>('closedBanners', null);

export const close = (bannerId: string): void => {
  let banners: string[] = get(closedBanners) ?? [];
  banners = [...banners, bannerId];
  banners = banners.slice(banners.length - 10, banners.length);

  closedBanners.set(banners);
};
