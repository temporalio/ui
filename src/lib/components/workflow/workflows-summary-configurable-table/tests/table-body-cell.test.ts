import { afterEach, beforeEach, describe, expect, test } from 'vitest';

import { WorkflowHeaderLabels } from '$lib/stores/workflow-table-columns';
import type { WorkflowExecution } from '$lib/types/workflows';

import TableBodyCell from '../table-body-cell.svelte';

const workflow: WorkflowExecution = {
  name: 'mock-workflow',
  id: 'abc-123',
  runId: 'def-456',
  startTime: '2023-04-27T00:00:00.000000000Z',
  endTime: '2023-04-28T00:00:00.000000000Z',
  executionTime: '2023-04-27T01:00:00.000000000Z',
  status: 'Running',
  taskQueue: 'task-queue',
  historyEvents: '42',
  historySizeBytes: '1024',
  parent: {
    workflowId: 'parent-workflow-id',
  },
  parentNamespaceId: 'parent-namespace-id',
  searchAttributes: {},
  pendingChildren: [],
  pendingActivities: [],
  stateTransitionCount: '12',
  url: 'the-url',
  isRunning: true,
  defaultWorkflowTaskTimeout: {},
  canBeTerminated: true,
};

let target: HTMLElement;

beforeEach(() => {
  target = document.createElement('div');
  target.setAttribute('id', 'target');
  document.body.appendChild(target);
});

afterEach(() => {
  target.remove();
});

describe(TableBodyCell.name, () => {
  test.each(WorkflowHeaderLabels)('%s renders', (label) => {
    const column = { label, pinned: false };
    const instance = new TableBodyCell({ target, props: { column, workflow } });
    expect(instance).toBeTruthy();
    expect(target.innerHTML).toMatchSnapshot();
  });
});
