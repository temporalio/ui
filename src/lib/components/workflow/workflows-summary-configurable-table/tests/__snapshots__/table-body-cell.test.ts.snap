// Vitest Snapshot v1

exports[`Table_body_cell$ > End renders 1`] = `"<td class=\\"workflows-summary-table-body-cell svelte-qupenl\\" data-testid=\\"workflows-summary-table-body-cell\\">2023-04-28 UTC 00:00:00.00</td>"`;

exports[`Table_body_cell$ > Execution Duration renders 1`] = `"<td class=\\"workflows-summary-table-body-cell svelte-qupenl\\" data-testid=\\"workflows-summary-table-body-cell\\">1 day</td>"`;

exports[`Table_body_cell$ > Execution Time renders 1`] = `"<td class=\\"workflows-summary-table-body-cell svelte-qupenl\\" data-testid=\\"workflows-summary-table-body-cell\\">2023-04-27 UTC 01:00:00.00</td>"`;

exports[`Table_body_cell$ > History Length renders 1`] = `"<td class=\\"workflows-summary-table-body-cell svelte-qupenl\\" data-testid=\\"workflows-summary-table-body-cell\\">42</td>"`;

exports[`Table_body_cell$ > History Size renders 1`] = `"<td class=\\"workflows-summary-table-body-cell svelte-qupenl\\" data-testid=\\"workflows-summary-table-body-cell\\">1 KB</td>"`;

exports[`Table_body_cell$ > Parent Namespace renders 1`] = `"<td class=\\"workflows-summary-table-body-cell svelte-qupenl\\" data-testid=\\"workflows-summary-table-body-cell\\">parent-namespace-id</td>"`;

exports[`Table_body_cell$ > Parent Workflow ID renders 1`] = `"<td class=\\"workflows-summary-table-body-cell svelte-qupenl\\" data-testid=\\"workflows-summary-table-body-cell\\">parent-workflow-id</td>"`;

exports[`Table_body_cell$ > Run ID renders 1`] = `"<td class=\\"workflows-summary-table-body-cell filterable svelte-qupenl\\" data-testid=\\"workflows-summary-table-body-cell\\"><a href=\\"/namespaces/default/workflows/abc-123/def-456/history\\" class=\\"table-link\\">def-456</a> </td>"`;

exports[`Table_body_cell$ > Start renders 1`] = `"<td class=\\"workflows-summary-table-body-cell svelte-qupenl\\" data-testid=\\"workflows-summary-table-body-cell\\">2023-04-27 UTC 00:00:00.00</td>"`;

exports[`Table_body_cell$ > State Transitions renders 1`] = `"<td class=\\"workflows-summary-table-body-cell svelte-qupenl\\" data-testid=\\"workflows-summary-table-body-cell\\">12</td>"`;

exports[`Table_body_cell$ > Status renders 1`] = `"<td class=\\"workflows-summary-table-body-cell svelte-qupenl\\" data-testid=\\"workflows-summary-table-body-cell\\"><span class=\\"flex text-center text-sm font-medium leading-4\\"><h6 class=\\"blue flex rounded-sm px-1 py-1 svelte-b7co3f\\"><span class=\\"whitespace-nowrap\\">Running</span> <div class=\\"heart-beat svelte-8dynez\\" style=\\"--animation-delay:0ms;\\"><div class=\\"heart-rate svelte-8dynez\\"><svg version=\\"1.0\\" xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" x=\\"0px\\" y=\\"0px\\" width=\\"30px\\" height=\\"18px\\" viewBox=\\"0 0 150 73\\" enable-background=\\"new 0 0 150 73\\" xml:space=\\"preserve\\"><polyline fill=\\"none\\" stroke=\\"#1D4ED8\\" stroke-width=\\"3\\" stroke-miterlimit=\\"10\\" points=\\"0,45.486 18.514,45.486 24.595,33.324 32.676,45.486 37.771,45.486 42.838,55.622 51.959,18 56.067,45 60.067,60.729 63.122,45.486 77.297,45.486 83.379,41.419 90.473,45.486 100,45.486\\"></polyline></svg> <div class=\\"fade-in svelte-8dynez\\"></div> <div class=\\"fade-out svelte-8dynez\\"></div></div> </div></h6></span></td>"`;

exports[`Table_body_cell$ > Task Queue renders 1`] = `"<td class=\\"workflows-summary-table-body-cell svelte-qupenl\\" data-testid=\\"workflows-summary-table-body-cell\\">task-queue</td>"`;

exports[`Table_body_cell$ > Type renders 1`] = `"<td class=\\"workflows-summary-table-body-cell filterable svelte-qupenl\\" data-testid=\\"workflows-summary-table-body-cell\\"><a href=\\"/namespaces/default/workflows/abc-123/def-456/history\\" class=\\"table-link\\">mock-workflow</a> </td>"`;

exports[`Table_body_cell$ > Workflow ID renders 1`] = `"<td class=\\"workflows-summary-table-body-cell filterable svelte-qupenl\\" data-testid=\\"workflows-summary-table-body-cell\\"><a href=\\"/namespaces/default/workflows/abc-123/def-456/history\\" class=\\"table-link\\">abc-123</a> </td>"`;
